### PR TITLE
Refactor util functions

### PR DESCRIPTION
**Pull request by 20lives**
following #2610 this is the first step at removing Lodash

### CODE REVIEW

**Comment by adamwathan** [comment]
Hey sorry, mind fixing the conflict then happy to merge? Can fix myself if needed.

**Comment by 20lives** [comment]
:wink:

**Comment by 20lives** [comment]
btw, I also ran additional tests where the experimental features array isn't empty

File: src/util/getAllConfigs.js [hunk 1]
```diff
@@ -1,21 +1,18 @@
 import defaultConfig from '../../stubs/defaultConfig.stub.js'
 import { flagEnabled } from '../featureFlags'
-import { flatMap, get } from 'lodash'
 
 export default function getAllConfigs(config) {
-  const configs = flatMap([...get(config, 'presets', [defaultConfig])].reverse(), (preset) => {
-    return getAllConfigs(preset)
-  })
+  const configs = (config && config.presets ? config.presets : [defaultConfig])
+    .reverse()
```
Comment (adamwathan) [inline src/util/getAllConfigs.js:6]: Should be careful here because `reverse` mutates the array, so we are directly mutating `config.presets` which could have side-effects outside of Tailwind itself.

Comment (20lives) [reply src/util/getAllConfigs.js:6]: Good catch.

File: src/util/getAllConfigs.js [hunk 2]
```diff
@@ -1,21 +1,19 @@
 import defaultConfig from '../../stubs/defaultConfig.stub.js'
 import { flagEnabled } from '../featureFlags'
-import { flatMap, get } from 'lodash'
 
 export default function getAllConfigs(config) {
-  const configs = flatMap([...get(config, 'presets', [defaultConfig])].reverse(), (preset) => {
-    return getAllConfigs(preset)
-  })
+  const configs = (config && config.presets ? config.presets : [defaultConfig])
```
Comment (adamwathan) [inline src/util/getAllConfigs.js:5]: Personally I find the reliance of truthiness and shortcircuiting stuff in these type of statements a bit obfuscating. We rely on `get` from lodash quite a lot throughout the codebase and I think making this sort of change everywhere is going to make the code quite a bit less declarative and harder to read.

I'd like some sort of abstraction around these sorts of "grab a key from an object but use this default if it's not there" operations which happen so much in the codebase. One very robust and very small library I've used for this in the past is `dlv`, perhaps we can just pull that in and start swapping these over?

https://github.com/developit/dlv

Comment (20lives) [reply src/util/getAllConfigs.js:5]: by combining es2020 **optional Chaining** and **nullish coalescing** you get something as declerative as get or dlv:
```JS
config?.presets ?? [defaultConfig]
```

Since you already have **Babel** configured, only need to update **eslint** ecmaVersion.


Anyhow it's your decision.

Comment (adamwathan) [reply src/util/getAllConfigs.js:5]: I'm good with that if you want to update the eslint config in this PR too then 👍🏻

**Comment by 20lives** [comment]
gone

Another conflict, I can't keep up :joy:

**Comment by 20lives** [comment]
Dropped `flattenColorPalette` commit, will add it in another PR.

**Comment by adamwathan** [comment]
Sorry another conflict 😬

**Comment by adamwathan** [comment]
Finally ;)

**Comment by 20lives** [comment]
Shocking, I can't even remember what I did here.
Was pretty sure this will not be relevant with all the new JIT stuff.
